### PR TITLE
Refactor: Improve JavaScript parser readability

### PR DESCRIPTION
**Pull request by fisker**
## Description

Especially the babel parser,  really hard to understand what happened inside `createParse`.

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/parser/try-combinations.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+"use strict";
+
+function tryCombinations(combinations, fn) {
+  let result;
+  const errors = [];
```
Comment (sosukesuzuki) [inline src/language-js/parser/try-combinations.js:5]: Does this need to be an array? If we only use the first one, I don't think it needs to be an array.

```js
function tryCombinations(combinations, fn) {
  let result;
  let error;
  for (const item of combinations) {
    try {
      result = fn(item);
      break;
    } catch (e) {
      if (!error) {
        error = e;
      }
    }
  }
  return [result, error];
}
```

Comment (fisker) [reply src/language-js/parser/try-combinations.js:5]: I thought babel throws the last error, I was going to use

```js
const [ast, ...errors] = tryCombinations
if(!ast) {
  throw getLast(errors)
}
```

but seems throw the first one passes the tests.

I think should be able to get all errors?

Comment (sosukesuzuki) [reply src/language-js/parser/try-combinations.js:5]: I think there is no problem in the case that the first one is thrown.

Comment (fisker) [reply src/language-js/parser/try-combinations.js]: Updated to return only first error https://github.com/prettier/prettier/pull/9783/commits/21a6183f400a087edfd0bfec71d0cfe3da5d5cb5

**Pull request merged**